Create DariaCagle2(obligatory).cpp

DariaCagle opened the pull request:
На будущее хочу узнать, как оформляются енамы. Я их оформила как константные переменные, но понимаю, что, скорее всего, это неправильно. Переделывать в этот раз не буду, но хотела бы узнать, как правильно (большими буквами или каким-то другим образом).

DariaCagle committed:
Create DariaCagle2(obligatory).cpp

MykolaSuperman commented on DariaCagle2(obligatory).cpp:
```diff
@@ -0,0 +1,173 @@
+  kSixty = 60,
+  kNinety = 90,
+  kThousand = 1000,
+  kZero = 0
```
enum чаще пишут большими буквами

MykolaSuperman commented on DariaCagle2(obligatory).cpp:
```diff
@@ -0,0 +1,173 @@
+  kNinety = 90,
+  kThousand = 1000,
+  kZero = 0
+};
```
мы еще enum как тип не изучали, но наперед забегая.
в enum добавляют не все возможные константы, а  добавляют константы которые между собой связаны, для вас это например 
```
  kOne = 1,
  kTwo,
  kThree,
  kFour,
  kFive,
  kSix,
```
это как группа свзязаных по смыслу констант
а не все в куче

DariaCagle replied:
Хорошо. Просто хотела попробовать)Можно в этом файле так оставить?

MykolaSuperman replied:
да только переименуйте имена для констант для switch оператора

MykolaSuperman commented on DariaCagle2(obligatory).cpp:
```diff
@@ -0,0 +1,173 @@
+  std::cout << "Enter any multidigital number:\n";
+  int number;
+  int i = kZero;
+  int sum = kZero;
```
ну это лишнее писать для инициализации перемнной нулем заводить константу. Константой заводят переменную которая по тексту например может встречаться несколько раз (например это размер массива), и ее значение (константы) например может измениться, потом константой выносят значение которое в выражении трудно понять (т.е. некий коэфициент  и т.д.)
В вашем случае просто инициализация нулем
```
int i{0};
или
int i{};
или 
int i = 0;
``` 
понятней намного

DariaCagle replied:
Принято. Это только с нулем?
Можно тоже так оставить?

MykolaSuperman replied:
можно только в следующих версиях, не пишите константу для литерала 0

MykolaSuperman commented on DariaCagle2(obligatory).cpp:
```diff
@@ -0,0 +1,173 @@
+  std::cout << "Enter any multidigital number:\n";
+  int number;
+  int i = kZero;
+  int sum = kZero;
```
ну это лишнее писать для инициализации перемнной нулем заводить константу. Константой заводят переменную которая по тексту например может встречаться несколько раз (например это размер массива), и ее значение (константы) например может измениться, потом константой выносят значение которое в выражении трудно понять (т.е. некий коэфициент  и т.д.)
В вашем случае просто инициализация нулем
```
int i{0};
или
int i{};
или 
int i = 0;
``` 
понятней намного

MykolaSuperman commented on DariaCagle2(obligatory).cpp:
```diff
@@ -0,0 +1,173 @@
+  while (number != kZero) {
+    digit = number % kTen;
+    sum += digit;
+    number = (number - (number % kTen)) / kTen;
```
смотрите описание выше
вместо `kTen` лучше писать 10. Смысл будет понятней чем заводить лишнюю константу

DariaCagle replied:
То есть, если константа по смыслу выражения никогда не изменится, ее лучше писать числом? В этом смысл? А та, которую разработчик может по ходу разработки изменить, записывается константой?

MykolaSuperman replied:
да это одно из идей. А вторая это когда в расчете формулы применяеться коэфициент и ему надо дать именованный смысл для понимания. Пример машином обучении есть коэфициенты кореляции модели например

MykolaSuperman commented on DariaCagle2(obligatory).cpp:
```diff
@@ -0,0 +1,173 @@
+  int number;
+  int i = kZero;
+  int sum = kZero;
+  int digit;
```
а зачем тут обьявлять эту переменную, почему ее не обьявить в while?

DariaCagle replied:
согласна. везде переделала

MykolaSuperman commented on DariaCagle2(obligatory).cpp:
```diff
@@ -0,0 +1,173 @@
+  std::cin >> number;
+  int number1 = (number - (number % kThousand)) / kThousand;
+  int number2 = number % kThousand;
+  int sum1 = kZero;
```
```suggestion
  int sum1 {0};
```

MykolaSuperman commented on DariaCagle2(obligatory).cpp:
```diff
@@ -0,0 +1,173 @@
+  int number1 = (number - (number % kThousand)) / kThousand;
+  int number2 = number % kThousand;
+  int sum1 = kZero;
+  int digit;
```
внесите переменную `digit` в скоуп цикла while

MykolaSuperman commented on DariaCagle2(obligatory).cpp:
```diff
@@ -0,0 +1,173 @@
+}
+
+void reverseNumbers() {
+  __int64 number;
```
тип __int64 является расширение компилятора, это не есть частью стандарта

MykolaSuperman commented on DariaCagle2(obligatory).cpp:
```diff
@@ -0,0 +1,173 @@
+
+void reverseNumbers() {
+  __int64 number;
+  std::cout << "Enter any 32x signed number:\n";
```
как это понять `32x signed number` 32х знаковое число?

DariaCagle replied:
Возможно, я неправильно поняла задание. Я поняла как 32-знаковое число как int32_t, то есть число signed int от -2,147,483,648 to 2,147,483,647? Или меня не туда унесло?) Я, в общем-то, это имела в виду/

MykolaSuperman replied:
можно было написать что `32 byte signed value`
но для пользователя ненужно такую информацию давать, он может вообще не понимать что такое byte и т.к. он понимает только значения

MykolaSuperman commented on DariaCagle2(obligatory).cpp:
```diff
@@ -0,0 +1,173 @@
+  std::cin >> number;
+  int digit;
+  __int64 reverseNumber = kZero;
+  if (number == abs(number)) {
```
я так понимаю это проверка на то что у вас отрицательное число или нет.
ну я думаю что этот код немного проще можно написать
```
bool is_negative = number < 0;

if (is_negative) {
  number = -number;
}
// ваш цикл

потом в конце
if (is_negative) {
  reverseNumber = - reverseNumber;
}
```

DariaCagle replied:
Не всегда получается написать то, что проще) Но это же не ошибка?

MykolaSuperman replied:
смотря что понимать под ошибкой. В вашем контексте вы усложняете чтения кода, т.е. вы добавили два цикла. Потом для человека записать вида `if (number == abs(number))` не так очевидно как if (number < 0).
Работающий код это не значит коректная программа. Нужно еще писать код как читающий код для других разработчиков. Потом в компиляторах есть такое понятие как оптимизация, и написав немного сумбурный код вы усложняете компилятору делать оптимизацию, но это пока не скоро вы узнаете и поймете)

MykolaSuperman commented on DariaCagle2(obligatory).cpp:
```diff
@@ -0,0 +1,173 @@
+  if (number == abs(number)) {
+    while (number != kZero) {
+      digit = number % kTen;
+      reverseNumber = reverseNumber * kTen + digit;
```
а если такая вот например будет ситуация
```
int8_t {144};
```
если сделать реверс это числа то получим `441`, и у нас будет наверное переполение?

DariaCagle replied:
Честно говоря, я не совсем понимаю, что вы имеете в виду. По моей задумке, эти переменные должны быть инт64, как раз для того, чтобы влезло любое 32-разрядное число. Опять же, если я правильно понимаю задание.

MykolaSuperman replied:
я не увидел что она у вас int64_t

MykolaSuperman commented on DariaCagle2(obligatory).cpp:
```diff
@@ -0,0 +1,173 @@
+void sumOfOdd() {
+  int numberOfElements;
+  do {
+    std::cout << "Enter number of elements from " << kOne << " to " << kFifty
```
Лучше константы `kOne` и `kFifty` обьявить внутри этой функции

MykolaSuperman commented on DariaCagle2(obligatory).cpp:
```diff
@@ -0,0 +1,173 @@
+              << ":\n";
+    std::cin >> numberOfElements;
+  } while ((numberOfElements < kOne) || (numberOfElements > kFifty));
+  int *elements = new int[numberOfElements];
```
нет, тут этого явно делать не нужно. Я бы не давал дз в котором нужно использовать то что мы явно не изучали.
Но важно то что эту задачу можно решить без массива, без выделения памяти.
Есть такое правило в программировании, хороший код это тот код в котором нет ничего лишнего, в вашем случае лишнее это динамический массив

DariaCagle replied:
Я знаю, как решить эту задачу без массива. Просто она почти такая же, как первая. Цикл с вводом элементов, количество которых запрашивается заранее, проверка на нечетность в на месте и приплюсование в сумму. Я просто хотела попробовать что-то новое. Тем более, оно же работает) Заверяю вас, я осознаю, что решение, которое легче, предпочтительнее, и осознаю, что решение через суммирование в цикле легче. Если хотите, могу переделать.

DariaCagle replied:
Хотя неважно, я уже переписала)

MykolaSuperman commented on DariaCagle2(obligatory).cpp:
```diff
@@ -0,0 +1,173 @@
+  int *elements = new int[numberOfElements];
+  std::cout << "Enter " << numberOfElements << " elements from " << -kSixty
+            << " to " << kNinety << ":\n";
+  size_t i = kZero;
```
```suggestion
  size_t i {};
```

MykolaSuperman commented on DariaCagle2(obligatory).cpp:
```diff
@@ -0,0 +1,173 @@
+    if (elements[count] % kTwo == kOne) {
+      sum = sum + elements[count];
+    }
+  }
```
два этих цикла модно написать одним циклом. подумайте

MykolaSuperman commented on DariaCagle2(obligatory).cpp:
```diff
@@ -0,0 +1,173 @@
+
+int sum(int number) {
+  int digit;
+  int sum = kZero;
```
```suggestion
  int sum {};
```

MykolaSuperman commented on DariaCagle2(obligatory).cpp:
```diff
@@ -0,0 +1,173 @@
+}
+
+int sum(int number) {
+  int digit;
```
обьявляйте переменную digit  внутри цикла

DariaCagle replied:
сделано

MykolaSuperman commented on DariaCagle2(obligatory).cpp:
```diff
@@ -0,0 +1,173 @@
+  std::cout << "Enter a positive number:\n";
+  int number;
+  std::cin >> number;
+  int bestDivisor = kZero;
+  for (int divisor = kOne; divisor <= number; divisor++) {
+    if ((number % divisor) == kZero) {
```
```suggestion
  int bestDivisor{0};
  for (size_t divisor {0}; divisor <= number; divisor++) {
    if ((number % divisor) == 0) {
```
в текущем контексте константа kZero является лишней.
Потом счетчик цикла всегда обьявляйте с типом `size_t`

DariaCagle replied:
Это я могу объяснить. Я делю инт намбер на дивайзор. Разве не будет компилятор ругаться, что здесь два разных типа? Я помню, вы что-то такое говорили.

MykolaSuperman replied:
а да не обратил внимание. Сорри. можете оставить как есть

MykolaSuperman commented on DariaCagle2(obligatory).cpp:
```diff
@@ -0,0 +1,173 @@
+    if ((number % divisor) == kZero) {
+      if ((sum(divisor) - bestDivisor) > kZero) {
+        bestDivisor = sum(divisor);
+      }
```
а зачем два раза вызывать функцию sum с одним и тем же аргументом?
напишите просто 
```
auto sum_digit_devisor = sum(divisor);
```
потом ваше условие 
```
if ((sum(divisor) - bestDivisor) > kZero)
```
немного только усложняет чтение вашего кода
```
if (sum_digit_devisor > bestDivisor) {
  bestDivisor = sum_digit_devisor;
}
```
 я думаю что читающему код будет понятней что вы тут делаете

DariaCagle replied:
сделано

MykolaSuperman commented on DariaCagle2(obligatory).cpp:
```diff
@@ -0,0 +1,173 @@
+    int chosenProgram;
+    std::cin >> chosenProgram;
+    switch (chosenProgram) {
+    case kOne:
```
вот в этом контексте константа в case ветке должна описывать что делает ветка
Как вы думаете если мне не читать то что вы вверху предлагате пользователю ввести, я пойму по константе  kOne что у вас вас будет делать ветка case?
Я думаю что лучше дать более понимаемое имя этой константе, ну например 
```
kOne -> kSumAverageTask;
kTwo -> kLuckyTicketTask;
и т.д.
```

DariaCagle replied:
полностью с вами согласна. только сегодня об этом думала. надо же давать константам говорящие имена, а не как я. а что тогда делать, если одна и та же константа встречается постоянно в коде, но в одном случае - это, как у меня, название кейса, а в другом просто число? еще одну константу вводить с говорящим именем? несколько констант с одинаковым значением, мне кажется, тоже не очень...

MykolaSuperman replied:
да если они разные по смыслу, то дают разные имена. Т.к. сейчас они могу быть с одинаковыми значениями, а завтра могут иметь разные значения, так их контекст абсолютно разный

MykolaSuperman commented on DariaCagle2(obligatory).cpp:
```diff
@@ -0,0 +1,173 @@
+      bestDivisor();
+      break;
+    default:
+      exit(1);
```
лучше return 0;
контекст применения exit немного другой

DariaCagle replied:
ок. удаляю

DariaCagle replied:
тогда же мне и стдлиб не нужен?

MykolaSuperman replied:
не нужен

DariaCagle committed:
Update DariaCagle2(obligatory).cpp

DariaCagle committed:
New Sum of Odd

DariaCagle committed:
Update DariaCagle2(obligatory).cpp

Добавила класс enum TASKS отдельно для кейсов заданий. Остальные енамы не буду переделывать, но я теперь поняла, где нужно енам, где константу, а где просто число. Вроде бы все.

MykolaSuperman commented on DariaCagle2(obligatory).cpp:
```diff
@@ -0,0 +1,204 @@
+    }
+  }
+  std::cout << sum << "\n";
+}*/
```
удаляет закоментированный код. Это плохая практика выкладывать на ревью код в котором есть закоментированные участки кода

DariaCagle replied:
удаляю

MykolaSuperman commented on DariaCagle2(obligatory).cpp:
```diff
@@ -0,0 +1,204 @@
+  } while ((numberOfElements < kOne) || (numberOfElements > kFifty));
+  std::cout << "Enter " << numberOfElements << " elements from " << -kSixty
+            << " to " << kNinety << ":\n";
+  int sum = 0;
```
опять же приучите себя писать 
```suggestion
  int sum{0};
```
если вы будете в привычке инициализировать переменные как вы делаете это сейчас, то в будущем у вас могут быть очень серьезные проблемы

DariaCagle replied:
я поняла и в следующих заданиях уже пишу с {}

MykolaSuperman commented on DariaCagle2(obligatory).cpp:
```diff
@@ -0,0 +1,204 @@
+      std::cout << "Wrong range! Enter again:\n";
+      std::cin >> element;
+    }
+    if (element % kTwo == kOne) {
```
еще раз если вы напишите вот так
```
if (element % 2 == 1)
```
код будет намного читабельней, не используйте константы там где они не нужны

DariaCagle replied:
с этим я тоже уже разобралась. в этот раз уже не хочу переделывать. я более-менее разобралась, какие переменные в каком виде писать)

DariaCagle committed:
Update DariaCagle2(obligatory).cpp

DariaCagle committed:
changed

Все поменяла. Можно проверять.

DariaCagle committed:
DariaCagle2 (optional)

DariaCagle committed:
Delete DariaCagle2(optional).cpp
